Reflection, Query Builders and Scope Creep

jongpie opened the pull request:
What started out as a journey to implement reflection in Nebula... turned into a journey of reflecting inward... and also, I decided to change a lot of stuff.

Fixes #19, fixes #36, fixes #37, fixes #50, fixes #52, fixes #68, fixes #69


Draft of release notes are below - I'll update them here as I wrap up the pull request.
_________________________

## NebulaFactory.cls
The framework now supports reflection for some interfaces. To utilise your own interface implementation, open the custom metadata type 'Nebula Interface Implementations', find the record of the desired interface and enter the name of your Apex class. Nebula will then inject your code into the framework at runtime. 
        1. IAggregateResultQueryBuilder
        2. IQueryArgumentFormatter
        3. IQueryFilter
        4. ISearchQueryBuilder
        5. ISObjectQueryBuilder

To leverage Apex's implementation of reflection, all relevant classes now use a parameterless constructor - any setup work needed has been moved from constructors to methods.

# SObject Repository
## DML.cls
* All DML methods now return the corresponding list of database results - see IDML.cls. Previously, these were void methods. This change is inherited by SObjectRepository.cls as well (it extends DML.cls)
## SObjectRepository.cls
* The member variable sobjectRepository.Query is now an instance of SObjectQueryBuilder.cls - see '*Query Builders*' section for more info
- New member variable sobjectRepository.AggregateQuery, an instance of the new class AggregateResultQueryBuilder.cls - see '*Query Builders*' section for more info

# Query Builders
QueryBuilder.cls has been split into SObjectQueryBuilder.cls, AggregateQueryBuilder.cls & SearchQueryBuilder.cls

## QueryBuilder.cls
* QueryBuilder.cls is now an abstract class. It only contains protected methods for shared logic that is used by SObjectQueryBuilder.cls & AggregateQueryBuilder.cls. QueryBuilder.cls itself should no longer be used directly.

## SObjectQueryBuilder.cls
A new builder class that generates dynamic SOQL queries & returns a list of SObject
* 7 ways to add fields to your query: 
        1. addFields(Schema.FieldSet fieldSet)
        2. addFields(List<Schema.SObjectField> fields)
        3. addAllFields()
        4. addAllStandardFields()
        5. addAllCustomFields()
        6. addAllReadableFields()
        7. addAllEditableFields()
* Name fields for lookups are automatically added to your query. For example, adding the contact object's field AccountId (a lookup to account):
    ```sobjectQueryBuilder.addField(Schema.Contact.AccountId)```

     will result in your query having 2 fields added 
     ```SELECT AccountId, Account.Name FROM Contact```
* New method for including children records
* New method, getQueryLocator()

## AggregateQueryBuilder.cls
A new builder class that generates dynamic SOQL queries & returns a list of AggregateResult
* Four methods to add fields to group by 
        1. groupBy(QueryDate queryDate);
        2. groupBy(Schema.FieldSet fieldSet);
        3. groupBy(Schema.SObjectField groupByField);
        4. groupBy(List<Schema.SObjectField> groupByFields);

* Six aggregate methods
        1. avg(Schema.SObjectField sbojectField)
        2. count(Schema.SObjectField sbojectField)
        3. count_distinct(Schema.SObjectField sbojectField)
        4. max(Schema.SObjectField sbojectField)
        5. min(Schema.SObjectField sbojectField)
        6. sum(Schema.SObjectField sbojectField)

Since this is the first round of implementing the class, not all aggregate result features (like 'GROUP BY ROLLUP') have been implemented, but more will be added in the future.

## SearchQueryBuilder.cls
A new builder class that generates dynamic SOSL queries & returns a list of a list of SObject (yes, a list of list, that's not a typo)
* Add SObject Query Builders to include them in the returned search results
        1.  addSObjectQueryBuilders(ISObjectQueryBuilder sobjectQueryBuilder) - add a single SObject query builder
        2.  addSObjectQueryBuilders(List<ISObjectQueryBuilder> sobjectQueryBuilders) - add a list of SObject query builders
* The query search group defaults to 'ALL FIELDS', but can be changed to any of the QuerySearchGroup enums
        1.  setQuerySearchGroup(QuerySearchGroup searchGroup)

## QueryFilter.cls
* Completely changed the methods available to handle 4 filtering scenarios
        1. filterByField
        2. filterByParentField
        3. filterByQueryDate
        4. filterBySubquery
* Added support for specifying 'AND' and 'OR' statements in a query. To use, provide a list of query filters that you want to be wrapped as a set of 'AND' or 'OR' statements to the corresponding method
        1. andFilterBy
        2. orFilterBy

## QueryDate.cls
* A new class that can be used in two ways
        1. Filter on a date/datetime field, using date functions
```
QueryDate qd = QueryDate.CALENDAR_MONTH(Schema.Lead.CreatedDate);
QueryFilter filter = new QueryFilter().filterByQueryDate(qd, QueryOperator.EQUALS, 3);
sobjectQueryBuilder.filterBy(filter);
System.assertEquals(filter.getValue(), 'CALENDAR_MONTH(CreatedDate) = 3');

// In SOQL, it would be written as
List<Lead> leads = [SELECT Id
FROM Lead
WHERE CALENDAR_MONTH(CreatedDate) = 3')];
```

   2. Group by a date function in an aggregate result

```
QueryDate qd = QueryDate.CALENDAR_MONTH(Schema.Lead.CreatedDate);
aggregateResultQueryBuilder.groupBy(qd);

// In SOQL, it would be written as
List<AggregateResult> results = [SELECT CALENDAR_MONTH(CreatedDate)
FROM Lead
GROUP BY CALENDAR_MONTH(CreatedDate)];
```

## QueryFilterScope.cls
* Added 3 new filter scope enums
        1. DELEGATED
        2. MY_TERRITORY - only works in orgs with territory management enabled
        3. MY_TEAM_TERRITORY - only works in orgs with territory management enabled

# Record Types
## SObjectRecordTypes.cls
* Now requires the abstract method `Schema.SObjectType getSObjectType()` to be implemented
# Settings
* The custom setting NebulaRepositorySettings__c.object has been renamed to NebulaSObjectQueryBuilderSettings__c.object

# Utilities
## CollectionUtils.cls
* Added 3 new methods for working with collections
        1. getLastItem
        2. pop
        3. splice

## Environment.cls
* New property, NamespacePrefix

# Documentation
* Started commenting classes & methods with [ApexDoc](https://github.com/SalesforceFoundation/ApexDoc). This will be an ongoing improvement.

jongpie committed:
Created NebulaFactory.cls to generate each interface's implementation

jongpie committed:
Added new custom metadata types for configuring the implementation

jongpie committed:
Commented out some incomplete code in NebulaFactory.cls

jongpie committed:
Converted SObjectRecordTypes.cls to have a parameterless constructor

jongpie committed:
Environment.cls now has the property NamespacePrefix

jongpie committed:
Converted DML.cls to have a parameterless constructor

jongpie committed:
Forgot to commit the new method added to ISObjectRecordTypes.cls

jongpie committed:
Made a pass a updating most of the classes to work with Apex's reflection

Several unit tests are still failing & some classes need more rework, but everything compiles now at least

jongpie committed:
Removed the ApexClass__c value from the custom metadata types

This should prevent overwriting any custom values when deploying new versions of Nebula. Null values (first install of Nebula) falls back to the default implementation

jongpie committed:
Temporarily committing some changes to QueryFilter.cls to help dependencies

jongpie committed:
Added a temp duplicated line of code to QueryBuilder until I can further refactor it

jongpie committed:
Moved the call to addCommonQueryFields() in QueryBuilder.cls

Since we don't get the SObject Type via the constructor anymore, we have to wait until we have it before we can add the common fields

jongpie committed:
Made more progress on converting QueryFilter.cls to use reflection

jongpie committed:
SObjectRecordTypes now uses the new version of NebulaFactory.getQueryFilter method

jongpie committed:
SObjectRepository now uses the new version of NebulaFactory.getQueryFilter method

jongpie committed:
Missed a line in SObjectRepository that used an old getQueryFilter method

jongpie committed:
Readded some needed methods to QueryFilter.cls

jongpie committed:
Fixed a method call in SObjectRepositoryMocks_Tests.cls

jongpie committed:
More reflection changes for QueryFilter.cls

jongpie committed:
QueryFilter.cls & its tests are now mostly working again

jongpie committed:
SObjectRecordTypes.cls was using the wrong SObject Type internally - fixed

jongpie committed:
Found a way to get the SObject Type for RecordType without a hardcoded string

jongpie committed:
Added some additional static methods to CollectionUtils.cls

jongpie committed:
Rewrote more of QueryFilter, simplified things where possible

It feels cleaner & better now, but I've been wrong before

jongpie committed:
Started cleaning up my mess

jongpie committed:
Removed an unneeded line from Logger.cls

jongpie committed:
Fixed & added more tests in QueryFilter_Tests.cls

jongpie committed:
w  h  i t    e s   p   a c e

jongpie committed:
Simplified the methods in NebulaFactory to all be parameterless, removed some unused ones

jongpie committed:
Switched to using serializePretty for a log statement in QueryBuilder.cls

jongpie committed:
Added an ugly but partially functional AggregateResultQueryBuilder.cls

QueryBuilder.cls is being divided into 2 classes - SObjectQueryBuilder.cls is next, then clean up and consolidation

jongpie committed:
Created AggregateResultQueryBuilder's equally ugly twin, SObjectQueryBuilder.cls

QueryBuilder.cls is being ripped apart and will eventually be abstract for shared query building logic

jongpie committed:
SObjectQueryBuilder.cls is so ugly, I forgot to add it to the last commit

jongpie committed:
Added a quick implementation for generating subqueries + added some TODOs

jongpie committed:
Added a copy of ApexDoc.jar to the repo to help with generating docs

jongpie committed:
Added .atom-build.yml to run some build steps

Currently just runs ApexDoc, but might expand the commands/tools in the future

jongpie committed:
Removed an unneeded line in apexdoc.bat

jongpie committed:
Updated .atom-build.yml to have the correct link for opening classes in github

jongpie committed:
Added some logic to automatically add the name field of the parent object when a lookup field is added

The name field could be Name, Subject, CaseNumber, etc depending on which sobject it is (name is the most common one)

jongpie committed:
QueryBuilder has been promoted to its own module

jongpie committed:
Added some new methods for adding fields to a query

jongpie committed:
Added a TODO to IQueryBuilder.cls

jongpie committed:
Added a TODO in SObjectQueryBuilder.cls

jongpie committed:
Made more progress on the query builder classes

jongpie committed:
Commenting out SObjectRepository_Tests.cls until the class is actually usable/testable

jongpie committed:
Commented out SObjectRecordTypes_Tests.cls until everything else is stable

jongpie committed:
Commenting out more tests because I'm the worst

I probably hopefully won't regret this

jongpie committed:
Added a lot of TODO items & apexdoc tags - to be finished later

jongpie committed:
Added some code to help prevent SOQL injections in QueryArgumentFormatter.cls

jongpie committed:
SObjectRecordTypes.cls now uses NebulaFactory.getSObjectQueryBuilder() again

jongpie committed:
Removed some unfinished & unrelated code for dynamic with/out sharing

Not even sure it would work, so removing for now. I might revisit in the future.

jongpie committed:
Cleaned up some access modifiers, variables, whitespace, etc

CodeClimate was upset about a few minor things

jongpie committed:
NebulaFactory now handles situations where there is no matching config

jongpie committed:
Started making SObjectQueryBuilder & AggregateResultQueryBuilder extend QueryBuilder

We're getting the band back together

jongpie committed:
Cleaned up a few bits in AggregateResultQueryBuilder.cls

jongpie committed:
Fixed a few other issues with AggregateResultQueryBuilder.cls

jongpie committed:
Added license header & standardised format of apexdoc in several classes

jongpie committed:
Added the basic logic to SObjectQueryBuilder.includeChildrenRecords()

jongpie committed:
Moved the method filterByWithSeparator to QueryBuilder.cls

jongpie committed:
Fixed some issues in SObjectQueryBuilder.includeChildrenRecords()

jongpie committed:
Cleaned up ISObjectQueryBuilder.cls

jongpie committed:
Added a few overloaded methods for inSubquery() & notInSubquery()

jongpie committed:
Cleaned up some of the query builder classes

jongpie committed:
Added another unit test for escaping single quotes (SOQL injection)

jongpie committed:
Removed a reference to SObjectQueryBuilder.getQuery() - method is now private

jongpie committed:
SObjectRepository.cls now uses NebulaFactory for both query builders

jongpie committed:
Consolidate more logic to be shared by QueryBuilder.cls

jongpie committed:
Deleted SObjectTypeDescriber.cls because it wasn't being used

jongpie committed:
Consolidated some ApexDoc groups

jongpie committed:
Fixed a small typo in the ApexDoc for CollectionUtils.isCollection()

jongpie committed:
Removed IQueryBuilder.cls interface since it currently has no public methods

jongpie committed:
Cleaned up some duplication in QueryBuilder.cls

jongpie committed:
Fixed some tests for AggregateResultQueryBuilder.cls

jongpie committed:
Commented out the count() method because it's weird

It can't be used with any grouping, etc, so I might not implement it

jongpie committed:
Created QueryDate.cls to handle group by date functions & filter date functions

Still a WIP though

jongpie committed:
Added another overloaded method to IQueryFilter.cls

jongpie committed:
Switched some properties in SObjectFieldDescriber.cls to be exposed as methods

jongpie committed:
Cleaned up naming convention for protected QueryBuilder methods & more

filterBy method has been replaced by andFilterBy() & orFilterBy()

jongpie committed:
Added QueryDate_Tests.cls, added ApexDoc info to QueryDate.cls

jongpie committed:
Added more unit tests to CollectionUtils_Tests.cls

jongpie committed:
Cleaned up some TODO items

jongpie committed:
Cleaned up more TODO items

jongpie committed:
Cleaned up & simplified the internals of QueryFilter.cls

jongpie committed:
Commented out part of SObjectRepositoryMocks.cls until I have time to revisit

jongpie committed:
Fixed a broken assert in CollectionUtils_Tests.cls

jongpie committed:
QueryFilter now handles subselects, instead of the query builder classes

jongpie committed:
Added unit tests to QueryFilter_Tests.cls

jongpie committed:
Fixed a failing unit test in SObjectQueryBuilder_Tests.cls

jongpie committed:
Removed @group-content from a few classes - I'll readd later if needed

jongpie committed:
More cleanup & ApexDoc added

jongpie committed:
QueryFilter now handles AND and OR statements

Query builder classes get to keep being stupid and just use IQueryFilter.getValue()

jongpie committed:
Added some javadoc to QueryFilter.cls

jongpie committed:
Added an extra space for legibility in QueryFilter.cls

jongpie committed:
Added a guard clause for null in QueryBuilder.doFilterBy()

jongpie committed:
Added some more ApexDoc

jongpie committed:
Cleaned up some old crap

jongpie committed:
QueryBuilder.cls now handles executing queriees & logging the results

This simplifies AggregateResultQueryBuilder.cls & SObjectQueryBuilder.cls a bit
Also fixed some inconsistent spacing/indentation

jongpie committed:
Updated CMT to match new classes & interfaces

jongpie committed:
Renamed NebulaRepositorySettings to NebulaSObjectQueryBuilderSettings

jongpie committed:
Added new tests to SObjectQueryBuilder_Tests.cls

Also fixed a few issues that came up while writing the tests

jongpie committed:
Added back NebulaLoggerSettings__c that was accidentally removed

jongpie committed:
Fixed a variable name in a comment

jongpie committed:
Removed an unnecessary variable in SObjectRecordTypes.cls

jongpie committed:
Small capitalization changes in Logger.cls

jongpie committed:
All DML methods now return the corresponding list of database results

jongpie committed:
Fixed some return types to match the new ones in IDML.cls

jongpie committed:
Removed ISObjectRepository from NebulaFactory

It isn't called within the framework, so no need to make it configurable at this point

jongpie committed:
Added ISObjectQueryBuilder.getQueryLocator()

jongpie committed:
Added some better formatting to some TODO items

jongpie commented:
@jamessimone I'm still working on wrapping up some changes for this pull request (and catching up on tests, release notes, etc), but if you're bored and want to take a look, this has most of the changes that I'm making for v2.0

jongpie committed:
Changed includeChildrenRecords to use a SObjectQueryBuilder as a parameter

Instead of passing individual pieces, it makes more sense to reuse the sobject query builder class

jongpie committed:
SOSL queries are now handled by SearchQueryBuilder.cls

jongpie commented:
I've added `SearchQueryBuilder.cls`, giving a total of 3 types of query builders now
  1. SearchQueryBuilder - SOSL queries
  2. SObjectQueryBuilder - standard SOQL queries
  3. AggregateResultQueryBuilder - SOQL queries for aggregate results

jongpie committed:
Added a step to sort sobject queries in SearchQueryBuilder.cls

This should keep the query generation consistent, regardless of the order that the sobject queries are added

jongpie committed:
Overloaded & renamed ISearchQueryBuilder.addSObjectQueryBuilders()

jongpie committed:
Updated a call to a renamed method in SObjectRepository.cls

jongpie committed:
Changed SObjectQueryBuilder.getSearchQuery() to handle no sobj options

In SOSL, you can specify 'RETURNING Account' with no options (vs 'RETURN Account(Id, Name WHERE Email != null) - the parentheses should only be added when the sobject type has 1 or more options set

jongpie committed:
Removed an extra '+' symbol

jongpie committed:
Fixed a bug with SObjectQueryBuilder.getChildQuery()

jamessimone commented:
@jongpie I began giving this an in-depth look yesterday, as I wanted to port the AggregateResultQueryBuilder over to our code.  I ended up taking a slightly different approach with a few of the design choices that you made - for example, I had the AggregateResultQueryBuilder extending QueryBuilder without a separate interface.  This made some of the method calls slightly clearer.  I also am not sure that I see the value in adding reflection for the QueryBuilder implementations - it's hard to imagine somebody wanting to override the default implementations (I would argue that the defaults are actually the reason somebody would want to use Nebula to begin with - nobody wants to have to re-write the query-level code!), and the work necessary to support reflection (specifically, moving the setting of the SObjectType for each QueryBuillder out of the constructors) ends up leading to additional hoops to jump through.  Does that make sense?

I also wrapped the implementation for grabbing the chained string for relationship names to another object outside of QueryFilter - the reason for this is that we have a concrete implementation right now that uses the orderBy call in QueryBuilder on a parent field (CustomObject__c.Lead__r.CreatedDate).  So I changed the implementation for IQueryBuilder's interface to accept this object, SObjectFieldWrapper, and that object handlers either a single SObjectField, or a chained list of related SObjectFields.  

This SObjectFieldWrapper I also used in the QueryFilter implementation so that there is only one "filterByField" method, so to speak; the first argument is replaced by the wrapper.  More than happy to drop that in a branch if you are interested in taking a look.

I need to look into the rest of the changes you made today, and I am sure you plan to make others as well.  Overall I have to say that the quality of the code is quite high, and aside from the stuff that I've mentioned (which is very specific in terms of feedback), it has been a pleasure to look through all of this!

jongpie commented:
@jamessimone your comments about reflection for the query builders totally makes sense - I started working on just implementing reflection, but query builder (the one class at the time) was giving me some problems when trying to convert to work with reflection, plus I knew I wanted to eventually make some enhancements to QueryBuilder.cls anyway, and well... I went down a rabbit hole of refactoring and changing. In the end, I think you're right - I don't think there's a benefit to making the query builder classes use reflection. I'm going to undo some of the reflection-related changes that I made and move some of the code back into the constructors. 

Your SObjectFieldWrapper class sounds like a great idea, that might actually help with a few other things that I want to add to the query builder classes in the future - if you're cool with throwing it into a branch, I'd love to take a look.

Last thing to note - I added SearchQueryBuilder.cls for SOSL queries. I'm not sure if you saw that change already or not, but it involved a few more changes to SObjectQueryBuilder.cls

jamessimone commented:
@jongpie I will throw that in a branch, then!  Also, two other notes: 
*  the getFirstQueryResult / getQueryResult calls in AggregateResultQueryBuilder can be eliminated entirely (!)
* I don't know if you will want to follow this approach, but I changed the aggregate methods in AggregateResultQueryBuilder to take in a string as well as the SObjectField, with the string being the alias.  I get what you were going for with the default names, and it would be nice to have that as a fallback (though with method overloading that's gonna grow the class by quite a bit) - we have production code already referencing distinct aliases in the aggregate results, and it was more convenient for me to just require an alias than it was to do any other thing.  Your thoughts?

jamessimone committed:
Sobject field wrapper (#71)

* SObjectFieldWrapper addition

* bumping metadata api version

* fix casing

jongpie committed:
Renamed SObjectFieldWrapper.cls to QueryField.cls

jongpie committed:
Tweaked a few bits in QueryField.cls to make it more consistent with the rest of Nebula

jongpie committed:
Converted code to use QueryField when possible

jongpie committed:
Forgot to commit AggregateResultQueryBuilder - it now uses QueryField too

jongpie committed:
Fixed some broken tests

jongpie committed:
SObjectRepository.cls now uses QueryField.cls

jongpie committed:
Forgot to change SObjectRepositoryMocks_Tests.cls to use QueryField.cls

jongpie commented:
@jamessimone couple of notes for you!
* My brain is fried, so I'm sure I'm overlooking something - why can the getFirstQueryResult / getQueryResult calls in AggregateResultQueryBuilder be eliminated?
* I definitely plan to add the ability to alias the aggregate functions, but I personally will end up using the auto-alias names, so I want to support both options. I originally started implementing 2 versions for each aggregate function (see below), but then I wanted to see if there was a cleaner way. So far, overloading still seems like the best option for supporting both
    * Auto-aliased: `avg(Schema.SObjectField numericSObjectField)`
    * Manually aliased: `avg(Schema.SObjectField numericSObjectField, String fieldAlias)`
* I've renamed SObjectFieldWrapper to QueryField (and for now, I'm sticking with your implementation of passing SObject fields in the constructor - seems to be the shortest code when using QueryField). It's now being used in these public methods (and probably a few other classes that I'm forgetting + test classes)
  * AggregateResultQueryBuilder - groupBy() and orderBy() now use QueryField
  * SObjectQueryBuilder - orderBy() now uses QueryField
  * QueryFilter - you changed this already for SObjectFieldWrapper, I just changed it to the new class name, QueryField 

**So, now we can group by, filter on and order by parent fields, which is pretty freakin' sweet. Thanks again for the help!**

I'm still cleaning up the code, catching up on tests and stripping out some of the reflection changes that I implemented, so still a bit more work for me to do on this PR, but keep the feedback coming if you have more ideas.

jongpie committed:
Converted some more bits to use QueryField. add field aliasing for aggregates

jongpie committed:
Fixed some broken tests

jongpie commented:
@jamessimone welp, I made 2 more changes...

1. I had previously forgotten to update the aggregate functions in AggregateQueryBuilder.cls to use QueryField instead of Schema.SObjectField. That's done, so now you could do something like this (my examples aren't fully working code, just snippets of the relevant bits, let me know if they're unclear though)
```
// Get the min created date of the parent object (account)
[SELECT MIN(Account.CreatedDate) FROM Contact];
```
by doing something like this
```
QueryField parentAcctCreatedDate = new QueryField(new List<SObjectField>(Contact.AccountId, Account.CreatedDate));
// Get the min created date of the parent object (account)
aggregateQueryBuilder.min(parentAcctCreatedDate);
```

2. All the aggregate functions for AggregateQueryBuilder have been overloaded so you can supply your own field alias - just for you, dawg. Example:
```
// Automatic field alias
aggregateResultQueryBuilder.avg(myQueryField);
// Bring Your Own Field Alias (BYOFA)
aggregateResultQueryBuilder.avg(myQueryField, 'wubbaLubbaDubDub');

jongpie committed:
Added SObjectRepository.BaseQueryFilters - all queries in the repo inherit filters added here

jongpie committed:
QueryBuilder.cls now sorts the list of strings for 'where clause' & 'order by'

jongpie committed:
QueryField & QueryFilter now implement Comparable so they can be sorted logically

jongpie committed:
QueryBuilder now dedupes any strings in whereClauseList & orderByList

jamessimone commented:
@jongpie due to the magical mysteries of Apex, you don't even need to cast a `List<SObject>` to a `List<AggregateResult>` - so your base `getQueryResults()` and `getFirstQueryResult()` methods can be called without needing specific methods to get the aggregate results.

ALSO - I ended up making the 'getQuery()' method on QueryBuilder virtual and added the following to the Aggregate's over-ridden implementation:
`if(this.aggregateFunctionList.isEmpty()) {'
            return super.getQuery();'
        `}`  (sorry there's some weird formatting there with the code snippet, but I'm sure you get the point)

That way if you have repositories overriding the default query builder, you can still do regular queries so long as you aren't doing anything with aggregates.

Grouping by parent fields is pretty awesome (as are those BYOA's) - thank YOU for adding those in!

jongpie committed:
Reverted most of the reflection changes

It overcomplicates things w/o any real benefit at the moment

jongpie committed:
Added IQueryField.cls

jongpie committed:
Converted everything to use IQueryField instead of QueryField

jongpie committed:
Fixed a test in SObjectQueryBuilder_Tests.cls

jongpie committed:
Removed the old page layout for the delete CMT

jongpie committed:
Added code to set currentModule on several query builder classes

jongpie committed:
Fixed a bug in QueryFilter

jongpie committed:
Renamed 2 methods in ISearchQueryBuilder.cls

jongpie committed:
Updated ref to a SearchQueryBuilder method in SObjectRepository

jongpie committed:
Added sorting to the child relationship queries in SObjectQueryBuilder.cls

This will keep the resulting queries consistent, regardless of the order that the child queries are added

jongpie committed:
Removed an unfinished method in SObjectQueryBuilder.cls

jongpie commented:
@jamessimone I think I'm at a stopping point now. Do you mind taking another glance at the changes? Test classes are still lacking, but the main classes should be good now. I've essentially removed all of the reflection bits, even though that was the original aim. Since I ended up refactoring so much other stuff and new adding query builders, etc, I don't think there's now much value in supporting reflection right now. It may get added back in the future, but I prefer using constructors for some parameters. 

I also want your opinion on QueryFilter in particular - it's still using a parameterless constructor, with separate methods for different filtering scenarios. I want to try to move back to passing parameters in the constructor - essentially, each 'filterby<something>' method would get converted to a constructor. Thoughts? Should I move back to parameters in the constructors? Or keep the methods as-is?

jamessimone commented:
@jongpie I will take a look!

jongpie committed:
Removed some methods in SearchQueryBuilder in favor of using constructor parms

jongpie committed:
Added a default COUNT(Id) function when no group by or aggreg are added

jongpie commented:
@jamessimone please don't hate me, but I just pushed one last set of changes for SearchQueryBuilder.cls. I realized this weekend that I want all 3 query builders to be usable as soon as they are constructed - meaning, as soon as you construct one, you should be able to call getQueryResults() and not receive a runtime error. In the real world, the query builders would also need filters added, query fields to be selected, etc but I think it's nice to have the base object be usable.

SearchQueryBuilder.cls didn't work this way - you had to call a few of the builder methods before the search query would be usable. I've changed the constructors in SearchQueryBuilder. As a result, SObjectRepository.cls no longer has the query builder variables (this.Query, this.AggregateResult and this.SearchResult) - instead, each method in SObjectRepo now just instantiates it's own query builder as needed.

AggregateResultQueryBuilder.cls had a small change added - if you run something like this (no 'group by' or aggregate functions specified), then a default 'COUNT(Id)' is added to the aggregate query.
`new AggregateResultQueryBuilder(Schema.Account.SObjectType).getQueryResults();`

jongpie committed:
Fixed some tests in SObjectRepository_Tests.cls

jongpie committed:
Committing a missing method in SearchQueryBuilder.cls

jongpie committed:
Fixed some duplicated code in SearchQueryBuilder.cls

jamessimone commented:
@jongpie no hate, and no worries!  Reviewing the rest of your changes (I am mostly caught up) was one of my plans this weekend - sorry I spent it going over Lightning stuff instead.  I plan to take a look and get this fully reviewed by the end of the week.

jamessimone commented on src/classes/SObjectRecordTypes.cls:
```diff
@@ -48,19 +48,19 @@ public abstract class SObjectRecordTypes extends NebulaCore implements ISObjectR
         if(cachedRecordTypesBySObjectMap.containsKey(this.sobjectName)) return;
 
-        // Trying to call Schema.RecordType.SObjectType confuses Apex, so we have to get it do an extra describe call
+        // Trying to call Schema.RecordType.SObjectType confuses Apex, so we have to get it through an extra describe call
```
that is kind of hilarious

jamessimone commented on src/classes/SObjectRepository.cls:
```diff
@@ -50,14 +27,14 @@ public abstract class SObjectRepository extends DML implements ISObjectRepositor
     // SOQL
     public virtual SObject getById(Id recordId) {
-        return this.Query
+        return new SObjectQueryBuilder(this.getSObjectType())
```
i really like this change!

jamessimone commented on src/classes/AggregateResultQueryBuilder.cls:
```diff
@@ -0,0 +1,188 @@
+        return queryString;
+    }
+
+    public AggregateResult getFirstQueryResult() {
```
did you try taking these out?  https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_calls_query_aggregateresult.htm - aggregate result is just a read-only SObject

jongpie replied:
Yo! Thanks for reviewing! I think I'm going to keep these methoda for now, just for the sake of clarity and being explicit when possible

jamessimone reviewed (approved):
Pretty much looks good man.  I think QueryDate, QueryField, and the new QueryBuilders are great additions.  I left the one comment about the aggregate results, but you may feel that having those methods there increases clarity.  Either way, thumb's up!

jongpie commented:
@jamessimone thanks for the review! I'm going to keep those methods for aggregate results for now. I plan to merge this to dev and master later today.
